driver/src/orderbook/streamed/order.rs

```diff
@@ -1,19 +1,106 @@
 use super::*;
+use crate::contracts::stablex_auction_element;
+use crate::models::Order as ModelOrder;
+use anyhow::Result;
 use serde::{Deserialize, Serialize};
 
+/// Change from a potential solution that might still get replaced by a better solution in the same
+/// batch.
+#[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+struct PendingUsedAmount {
+    batch_id: BatchId,
+    amount: u128,
+}
+
 #[derive(Copy, Clone, Debug, Deserialize, Serialize)]
 pub struct Order {
     pub buy_token: TokenId,
     pub sell_token: TokenId,
-    pub valid_from: BatchId,
     pub valid_until: BatchId,
-    pub price_numerator: u128,
-    pub price_denominator: u128,
-    pub used_amount: u128,
+    pub valid_from: BatchId,
+    price_numerator: u128,
+    price_denominator: u128,
+    // Invariant: used_amount + pending_used_amount.amount <= price_denominator
+    used_amount: u128,
+    pending_used_amount: PendingUsedAmount,
 }
 
 impl Order {
+    pub fn new(
+        buy_token: TokenId,
+        sell_token: TokenId,
+        valid_from: BatchId,
+        valid_until: BatchId,
+        price_numerator: u128,
+        price_denominator: u128,
+    ) -> Self {
+        Self {
+            buy_token,
+            sell_token,
+            valid_from,
+            valid_until,
+            price_numerator,
+            price_denominator,
+            used_amount: 0,
+            pending_used_amount: PendingUsedAmount {
+                batch_id: 0,
+                amount: 0,
+            },
+        }
+    }
+
     pub fn has_limited_amount(&self) -> bool {
         self.price_numerator != std::u128::MAX && self.price_denominator != std::u128::MAX
     }
+
+    pub fn is_valid_in_batch(&self, batch_id: BatchId) -> bool {
+        self.valid_from <= batch_id && batch_id <= self.valid_until
+    }
+
+    pub fn get_used_amount(&self, batch_id: BatchId) -> u128 {
+        if self.pending_used_amount.batch_id < batch_id {
+            // Cannot fail because of the invariant.
+            self.used_amount
+                .checked_add(self.pending_used_amount.amount)
+                .unwrap()
+        } else {
+            self.used_amount
+        }
+    }
+
+    pub fn get_remaining_amount(&self, batch_id: BatchId) -> u128 {
```

Comment:
I think if the account state is only ever constructed with events that are to be considered for a batch, then passing the batch ID here shouldn't be necessary.

Reply:
If I remember correctly we dismissed the idea of instantiating the orderbook reader with a target batch and only processing the events that matter for that batch.
Otherwise the same would apply for balances as you would know upfront if a deposit will still be pending or already matured.

Reply:
Ahh! I guess I misunderstood a bit what we had decided. I had understood that we didn't want specialized logic for filtering events based on their data, but that we would still construct the orderbook only based on the events up to a certain block number (determined based on the batch ID that we were fetching for).

The only tiny nit I have with the API then is that calling `order.get_remaining_amount(some_batch_in_the_past)` doesn't actually return the amount at the batch that was specified but the current batch.

Reply:
It returns the current remaining amount but to know whether pending is part of current it needs to know the batch id. Looking at this function in isolation I can see that it might be confusing but it's done like this throughout `State`.

Reply:
>  Looking at this function in isolation I can see that it might be confusing

I don't see a way around it though. I 100% OK with how it is right now.